File: bindings/deno/src/safe_email.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeEmail - Email validation that cannot crash.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe email operations backed by formally verified Idris 2 code.
 */
export class SafeEmail {
  /**
   * Validate an email address (RFC 5321 simplified).
   *
   * @param email - The email address to validate.
   * @returns Result containing a boolean or an error string.
   */
  static isValid(email: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(email);
    const result = symbols.proven_email_is_valid(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }
}

/**
 * Parsed email representation.
 * Created via SafeEmail.isValid() check then string splitting.
 */
export class Email {
  #raw: string;

  constructor(raw: string) {
    this.#raw = raw;
  }

  /**
   * Parse and validate an email string.
   *
   * @param email - The email to parse.
   * @returns Result containing an Email instance or an error string.
   */
  static parse(email: string): Result<Email> {
    const valid = SafeEmail.isValid(email);
    if (!valid.ok) return valid;
    if (!valid.value) return err('Invalid email address');
    return ok(new Email(email));
  }

  /** The full email address. */
  toString(): string {
    return this.#raw;
  }
}

File: bindings/ephapax-affine/src/safe_email.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Email validation with affine type semantics.
//!
//! All validation is performed by libproven's formally verified Idris 2
//! core. This module provides affine-typed wrappers that ensure validation
//! results are consumed at most once.

use crate::ffi;
use crate::{bool_result_to_proven, ProvenResult};

/// Affine wrapper for a validated email address.
///
/// `#[affine]` -- this value can be consumed at most once.
/// Once the email is extracted, the validation proof is consumed.
/// This ensures each validated email is used exactly once in
/// security-critical contexts.
#[derive(Debug)]
pub struct ValidatedEmail {
    address: String,
}

impl ValidatedEmail {
    /// Consume this validated email, returning the address string.
    ///
    /// After calling this, the `ValidatedEmail` is consumed and cannot be reused.
    pub fn consume(self) -> String {
        self.address
    }

    /// Borrow the email address without consuming the validation proof.
    pub fn as_str(&self) -> &str {
        &self.address
    }
}

/// Validate an email address (RFC 5321/5322).
///
/// Returns `Ok(true)` if the email is valid, `Ok(false)` if invalid.
/// Delegates to `proven_validate_email` via FFI.
pub fn is_valid_email(email: &str) -> ProvenResult<bool> {
    let bytes = email.as_bytes();
    // SAFETY: bytes pointer is valid for bytes.len() bytes.
    let result = unsafe {
        ffi::proven_validate_email(bytes.as_ptr(), bytes.len())
    };
    bool_result_to_proven(result)
}

/// Validate an email address and return a `ValidatedEmail` on success.
///
/// Returns `Ok(ValidatedEmail)` if the address is valid, or
/// `Err(ProvenError)` if the address is invalid or FFI fails.
/// Delegates to `proven_validate_email` via FFI.
///
/// The `ValidatedEmail` wrapper proves that validation occurred and
/// can be consumed at most once (affine semantics).
pub fn validate_email(email: &str) -> ProvenResult<ValidatedEmail> {
    let bytes = email.as_bytes();
    // SAFETY: bytes pointer is valid for bytes.len() bytes.
    let result = unsafe {
        ffi::proven_validate_email(bytes.as_ptr(), bytes.len())
    };
    let valid = bool_result_to_proven(result)?;
    if valid {
        Ok(ValidatedEmail {
            address: email.to_owned(),
        })
    } else {
        Err(crate::ProvenError {
            code: crate::STATUS_ERR_VALIDATION_FAILED,
            message: "Email validation failed",
        })
    }
}
